test/test_reductions.py

```diff
@@ -814,6 +814,17 @@ def test_prod(self, device, dtype):
         torch.prod(x, 1, out=res2)
         self.assertEqual(res1, res2)
 
+    def test_prod_bool(self, device):
+        tests = [
+            ([True, True], True),
+            ([True, False], False),
+            ([False, False], False)
+        ]
+
+        for val, exp in tests:
+            tensor = torch.tensor(val, device=device)
+            self.assertEqual(torch.prod(tensor, dtype=torch.bool), torch.tensor(exp, device=device))
```

Comment:
What happens if dtype isn't set? That is, what is the default behavior when a bool tensor is passed to prod?

Reply:
In this case (a bool tensor passed to prod without `dtype` variate), an `int64` type Tensor would be returned, which is consistent with the performance in NumPy.
```
>>> np.prod(np.array([True,True]))
1
>>> torch.prod(torch.tensor([True, True]))
tensor(1)
>>> np.prod(np.array([True,True])).dtype
dtype('int64')
>>> torch.prod(torch.tensor([True, True])).dtype
torch.int64
```
The only difference is that for the input of an empty Tensor, the result type of PyTorch is `float32`, while for NumPy, it is `float64`. I think it's caused by the different default types between PyTorch and NumPy, so I guess it's just fine.
```
>>> np.prod(np.array([]))
1.0
>>> np.prod(np.array([])).dtype
dtype('float64')
>>> torch.prod(torch.tensor([]))
tensor(1.)
>>> torch.prod(torch.tensor([])).dtype
torch.float32
```

Reply:
If the returned dtype is int64 when a boolean tensor is given as input, which code path is exercised in the function for boolean tensors? Because won't `iter.dtype()` return intt64?

Reply:
I think the `iter.dtype()` here would return `int64`. The code would finaly goes to `prod_out_impl`, and it would use `get_dtype` method to casting inputs from `kBool` type to `kLong` before the Iterator be created.
https://github.com/pytorch/pytorch/blob/a20d4511e40264a900c62ff9631fc063c40ef41e/aten/src/ATen/native/ReduceOps.cpp#L426-L430
https://github.com/pytorch/pytorch/blob/a20d4511e40264a900c62ff9631fc063c40ef41e/aten/src/ATen/native/ReduceOps.cpp#L489-L492

Reply:
That's my thinking, too. Would you verify? Also, if that's the case, then is this code ever used?

```
if (iter.dtype() == ScalarType::Bool) {
  using scalar_t = bool;
  binary_kernel_reduce_vec(
    iter,
    [=](scalar_t a, scalar_t b) -> scalar_t { return a && b; },
    [=](Vec256<scalar_t> a, Vec256<scalar_t> b) { return a && b; },
    /*identity=*/1);
```

Reply:
Yep, As shown below, I verified it in my environment. 

When setting dtype as `torch.bool`, it would using the code of `a&&b` , and in other cases (including the above-mentioned default `int64` scene), the part of `a*b` would be used. 

Such verification has been verified in both CPU and CUDA environments, and the performance of the two is consistent.
```
static void prod_kernel_impl(TensorIterator& iter) {
  // Workaround for the error: '*' in boolean context, suggest '&&' instead [-Werror=int-in-bool    -context]
  if (iter.dtype() == ScalarType::Bool) {
    std::cout << "Using a && b" << std::endl;
    using scalar_t = bool;
    binary_kernel_reduce_vec(
      iter,
      [=](scalar_t a, scalar_t b) -> scalar_t { return a && b; },
      [=](Vec256<scalar_t> a, Vec256<scalar_t> b) { return a && b; },
      /*identity=*/1);
  } else {
    std::cout << "Using a * b" << std::endl;
    AT_DISPATCH_ALL_TYPES_AND_COMPLEX(iter.dtype(), "prod_cpu", [&] {
      binary_kernel_reduce_vec(
        iter,
        [=](scalar_t a, scalar_t b) -> scalar_t { return a * b; },
        [=](Vec256 <scalar_t> a, Vec256 <scalar_t> b) { return a * b; },
        /*identity=*/1);
      });
  }
}
```
```
>>> import torch
>>> torch.prod(torch.tensor([True, True]), dtype=torch.bool)
Using a && b
tensor(True)
>>> torch.prod(torch.tensor([True, True]))
Using a * b
tensor(1)
>>> torch.prod(torch.tensor([True, True], device='cuda'), dtype=torch.bool)
Using a && b
tensor(True, device='cuda:0')
>>> torch.prod(torch.tensor([True, True]))
Using a * b
tensor(1)
```

Reply:
Thank you for the explanation, @Kiyosora, that's very helpful.

In the future we may want to become more clever with this dispatch so we can compute in bool when the input is bool, even if the output dtype is something else (like int64), but this seems OK for now.

I've made one suggestion about extending the test to validate both the default bool->int64 behavior is consistent with NumPy and the already tested bool->bool behavior is, too.